pkl-gradle/src/main/java/org/pkl/gradle/task/EvalTask.java

```diff
@@ -16,50 +16,69 @@
 package org.pkl.gradle.task;
 
 import java.io.File;
+import java.util.Set;
 import org.gradle.api.file.DirectoryProperty;
 import org.gradle.api.file.RegularFileProperty;
 import org.gradle.api.provider.Property;
+import org.gradle.api.provider.Provider;
 import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.Internal;
 import org.gradle.api.tasks.Optional;
-import org.gradle.api.tasks.OutputDirectory;
-import org.gradle.api.tasks.OutputFile;
+import org.gradle.api.tasks.OutputDirectories;
+import org.gradle.api.tasks.OutputFiles;
 import org.pkl.cli.CliEvaluator;
 import org.pkl.cli.CliEvaluatorOptions;
 
 public abstract class EvalTask extends ModulesTask {
-  @OutputFile
-  @Optional
+  @Internal
   public abstract RegularFileProperty getOutputFile();
 
-  @Input
-  @Optional
+  @Internal
   public abstract Property<String> getOutputFormat();
 
   @Input
   @Optional
   public abstract Property<String> getModuleOutputSeparator();
 
-  @OutputDirectory
-  @Optional
+  @Internal
   public abstract DirectoryProperty getMultipleFileOutputDir();
 
   @Input
   @Optional
   public abstract Property<String> getExpression();
 
+  @Internal
+  public Provider<CliEvaluator> getCliEvaluator() {
```

Comment:
I don't think `Provider` helps for this and the following methods, as they are only called if and when the information is needed.

Reply:
> Now the only question is if this works for a provider obtained with getProject().provider(). Another option is to inject a ProviderFactory into the task.

Hm... probably not, in that case. I'm guessing we need to reference the inputs.

Reply:
IIRC `getProject` would also fail if using configuration cache.
See https://docs.gradle.org/current/userguide/configuration_cache.html#config_cache:requirements:use_project_during_execution

Reply:
Good to know. I think that `@Inject`ing a `ProviderFactory` might be the right solution. Could check task dependency tracking with an automated or manual test.

Reply:
Related: https://discuss.gradle.org/t/auto-create-task-dependecies-in-case-of-computed-outputfile-provider-regularfile/45381